### ISSUE TITLE
Support for captive portals

### ISSUE THREAD

**Issue by brbsix** [thp/urlwatch#104]
Lately my internet connection has been behind a [captive portal](https://www.chromium.org/chromium-os/chromiumos-design-docs/network-portal-detection). Every so often all web requests are redirected to a landing page in order for me to sign in again. As you can imagine, this causes issues with urlwatch. The url jobs think they are getting a valid response when they are not.

I wonder if anyone else has had similar issues? I think the following will work for my purposes:

```
diff --git a/lib/urlwatch/jobs.py b/lib/urlwatch/jobs.py
index d268ef2..1d03517 100644
--- a/lib/urlwatch/jobs.py
+++ b/lib/urlwatch/jobs.py
@@ -61,6 +61,10 @@ class ShellError(Exception):
     def __str__(self):
         return '%s: Exit status %d' % (self.__class__.__name__, self.result)

+class CaptivePortalError(Exception):
+    """Exception raised when behind a captive portal"""
+    ...
+
 class NotModifiedError(Exception):
     """Exception raised on HTTP 304 responses"""
     ...
@@ -191,6 +195,9 @@ class UrlJob(Job):
         return self.url

     def retrieve(self, job_state):
+        if requests.get('http://clients3.google.com/generate_204').status_code == 200:
+            raise CaptivePortalError()
+
         headers = {
             'User-agent': urlwatch.__user_agent__,
         }
```

Obviously that's not going to be desirable for most people, but perhaps it makes sense to add some sort of optional configuration?

**Comment by thp** [comment]
Well, I guess that makes sense -- can you turn this into a pull request?
Is that URL publicly accessible in the sense that we can use it for captive portal usage and is not just some internal URL that Google uses for its apps?
Does it make sense to make the URL configurable?

**Comment by jwilk** [comment]
Some captive portals are implemented via HTTP redirects.
`!= 204` instead of `== 200` would handle this case.

Please make the feature disabled by default.
I don't want urlwatch to contact any sites I didn't explicitly ask it to.

**Comment by sbraz** [comment]
@brbsix Can't you just apply a custom filter to all URLs and make it throw a `NotModifiedError'? That's what I do for sites who have a tendency to fail.

**Comment by thp** [comment]
Since the captive portal issue seems to be something that only happens on a mobile computer (assuming more stationary computers and servers will never see a captive portal), maybe it makes sense to handle this in a wrapper script?

The page https://captivebehavior.wballiance.com lists some mechanisms that different clients use to check if connectivity works, it should be easy to put one of them into a small Python script that does the captive portal check, and if there's no captive portal, run `urlwatch` (and if there is a captive portal, either just not run `urlwatch` or send some kind of notification, etc..).

As @jwilk mentioned, hardcoding a specific captive portal isn't a good idea, and for most users I'd assume that this isn't needed, as the device where urlwatch runs is usually not a mobile computer (which isn't always on and not always connected).

tl;dr: If you are affected by captive portals, don't run `urlwatch` directly, but create/use a script that checks if there's a captive portal, and only calls `urlwatch` if the connection is working.
